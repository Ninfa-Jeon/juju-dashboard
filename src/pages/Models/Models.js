import React from "react";
import { useSelector } from "react-redux";
import { useLocation } from "react-router-dom";
import queryString from "query-string";

import Layout from "components/Layout/Layout";
import Header from "components/Header/Header";
import ModelTableList from "components/ModelTableList/ModelTableList";
import ButtonGroup from "components/ButtonGroup/ButtonGroup";

import SearchAndFilter from "@canonical/react-components/dist/components/SearchAndFilter";

import useWindowTitle from "hooks/useWindowTitle";

import { useQueryParam, StringParam, withDefault } from "use-query-params";

import { getGroupedModelStatusCounts } from "app/selectors";
import { pluralize } from "app/utils";

import "./_models.scss";

export default function Models() {
  useWindowTitle("Models");

  const [groupModelsBy, setGroupModelsBy] = useQueryParam(
    "groupedby",
    withDefault(StringParam, "status")
  );

  // Grab filter from 'groupedby' query in URL and assign to variable
  const location = useLocation();
  const queryStrings = queryString.parse(location.search, {
    arrayFormat: "comma",
  });
  let activeFilters = queryStrings.activeFilters;
  if (typeof activeFilters === "string") {
    // Maintain a consistent type returned from the parsing of the querystring.
    activeFilters = [activeFilters];
  }

  const { blocked, alert, running } = useSelector(getGroupedModelStatusCounts);
  const models = blocked + alert + running;

  return (
    <Layout>
      <Header>
        <div className="models__header">
          <div className="models__count">
            {`${models} ${pluralize(
              models,
              "model"
            )}: ${blocked} blocked, ${alert} ${pluralize(
              alert,
              "alert"
            )}, ${running} running`}
          </div>
          <ButtonGroup
            activeButton={groupModelsBy}
            buttons={["status", "cloud", "owner"]}
            label="Group by:"
            setActiveButton={setGroupModelsBy}
          />
          <SearchAndFilter
            filterPanelData={[
              {
                id: 0,
                heading: "Cloud",
                chips: [
                  { lead: "Cloud", value: "Google" },
                  { lead: "Cloud", value: "AWS" },
                  { lead: "Cloud", value: "Azure" },
                ],
              },
              {
                id: 1,
                heading: "Region",
                chips: [
                  { lead: "Region", value: "us-east1" },
                  { lead: "Region", value: "us-north2" },
                  { lead: "Region", value: "us-south3" },
                  { lead: "Region", value: "us-north4" },
                  { lead: "Region", value: "us-east5" },
                  { lead: "Region", value: "us-south6" },
                  { lead: "Region", value: "us-east7" },
                  { lead: "Region", value: "us-east8" },
                  { lead: "Region", value: "us-east9" },
                  { lead: "Region", value: "us-east10" },
                ],
              },
              {
                id: 2,
                heading: "Owner",
                chips: [
                  { lead: "Owner", value: "foo" },
                  { lead: "Owner", value: "bar" },
                  { lead: "Owner", value: "baz" },
                ],
              },
            ]}
            returnSearchData={(searchData) => {
              if (searchData.length > 0) {
                // do something
              }
            }}
          />
        </div>
      </Header>

      <div className="l-content">
        <div className="models">
          <ModelTableList groupedBy={groupModelsBy} filters={activeFilters} />
        </div>
      </div>
    </Layout>
  );
}
