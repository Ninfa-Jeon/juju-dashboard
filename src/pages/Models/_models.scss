.models {
  &__count {
    font-size: 0.875rem;
    padding: 0 1.5rem;
  }

  &__header {
    align-items: center;
    display: grid;
    grid-template-columns: repeat(3, 1fr) 3rem;
  }

  tr {
    display: grid;
    grid-template-columns: 1fr 0.6fr 0.75fr 1.15fr 1fr 0.5fr 0.65fr;
  }

  .cloud-group tr {
    grid-template-columns: 0.8fr 0.75fr 0.6fr 0.75fr 1.15fr 1fr 0.5fr 0.65fr;
  }

  // Tooltips override
  .p-tooltip--top-center .p-tooltip__message {
    left: calc(50% - 0.875rem);
    margin-bottom: -0.5rem;
  }
}
