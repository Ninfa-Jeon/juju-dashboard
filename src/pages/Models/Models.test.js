import React from "react";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router";
import { BrowserRouter as Router } from "react-router-dom";
import dataDump from "testing/complete-redux-store-dump";

import Models from "./Models";

const mockStore = configureStore([]);

describe("Models page", () => {
  it("renders without crashing", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Models />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("Header"));
    expect(wrapper.find("TableList"));
  });

  it("has a header which shows the model counts", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Models />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find(".models__count").text()).toBe(
      "17 models: 2 blocked, 4 alerts, 11 running"
    );
  });

  it("displays correct grouping view", () => {
    global.window = { location: { pathname: "/models?groupedby=owner" } };
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <Models />
        </Router>
      </Provider>
    );
    expect(
      wrapper.find(".p-model-group-toggle__button.is-selected").text()
    ).toBe("status");
    wrapper.find("button[value='owner']").simulate("click", {
      target: { value: "owner" }
    });
    wrapper.update();
    expect(
      wrapper.find(".p-model-group-toggle__button.is-selected").text()
    ).toBe("owner");
    expect(global.window.location.search).toEqual("?groupedby=owner");
    expect(wrapper.find(".owners-group"));
  });
});
