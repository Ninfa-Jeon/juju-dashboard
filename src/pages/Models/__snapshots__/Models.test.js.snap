// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Models page renders without crashing and matches snapshot 1`] = `
<Layout
  sidebar={true}
>
  <div
    className="p-strip is-shallow"
  >
    <div
      className="row"
    >
      <TableList />
    </div>
  </div>
</Layout>
`;
